tools/ot/src/main/java/dk/alexandra/fresco/tools/ot/otextension/BristolOtSender.java

```diff
@@ -0,0 +1,111 @@
+  private Pair<List<StrictBitVector>, List<StrictBitVector>> randomMessages;
+  // Index of the current random OT to use
+  private int offset = -1;
+
+  /**
+   * Initializes the underlying Rot functionality using {@code rotSender}. It
+   * will then construct OTs in batches of {@code batchSize}.
+   *
+   * @param rotSender
+   *          The underlying random OT sender object to use
+   * @param resources
+   *          The resource pool for this specific instance
+   * @param network
+   *          The network to use
+   * @param batchSize
+   *          Size of the OT extension batch the protocol will construct
+   */
+  public BristolOtSender(RotSender rotSender, OtExtensionResourcePool resources,
+      Network network, int batchSize) {
+    this.sender = rotSender;
+    this.resources = resources;
+    this.network = network;
+    this.batchSize = batchSize;
+  }
+
+  /**
+   * Send the serialized message from the current 1-out-of-2 OT.
+   *
+   * @param messageZero
+   *          The message to send for choice zero
+   * @param messageOne
+   *          The message to send for choice one
+   */
+  public void send(byte[] messageZero, byte[] messageOne) {
+    // Check if there is still an unused random OT stored, if not, execute a
+    // random OT extension
+    if (offset < 0 || offset >= batchSize) {
+      randomMessages = sender.extend(batchSize);
+      offset = 0;
+    }
+    doActualSend(messageZero, messageOne);
+    offset++;
+  }
+
+  /**
+   * Adjust the random, preprocessed message, to fit the specific messages to
+   * send.
+   *
+   * @param messageZero
+   *          The actual zero message to send
+   * @param messageOne
+   *          The actual one message to send
+   */
+  private void doActualSend(byte[] messageZero, byte[] messageOne) {
+    // Find the correct preprocessed random OT messages
+    StrictBitVector randomZero = randomMessages.getFirst().get(offset);
+    StrictBitVector randomOne = randomMessages.getSecond().get(offset);
+    int maxLength = Math.max(messageZero.length, messageOne.length);
+    // Receive a bit from the receiver indicating whether the zero and one
+    // messages should be switched around
+    byte[] switchBit = network.receive(resources.getOtherId());
+    // If false (indicated by byte 0x00), then don't switch around
+    if (switchBit[0] == 0x00) {
+      sendAdjustedMessage(messageZero, maxLength, randomZero.toByteArray());
+      sendAdjustedMessage(messageOne, maxLength, randomOne.toByteArray());
+    } else {
+      sendAdjustedMessage(messageOne, maxLength, randomZero.toByteArray());
+      sendAdjustedMessage(messageZero, maxLength, randomOne.toByteArray());
+    }
+  }
+
+  private void sendAdjustedMessage(byte[] realMessage, int maxLength,
+      byte[] randomMessage) {
+    byte[] toSend = new byte[maxLength];
+    // Use the random message as a a seed for a PRG
+    Drbg currentPrg = new AesCtrDrbg(randomMessage);
```

Comment:
As we talked about this can be optimized. We should consider the following:
- When the length of `randomMessage` is longer or equal `realMessage` we should not use a PRG and simply use `randomMessage` directly for the pad.
- In all other cases, there is no reason to initialize a new Drbg on each call. Instead use a hash function to stretch the random message.

Reply:
This is part of task https://github.com/aicis/fresco/issues/218

Reply:
I guess we can do it after we merge this. It seems like very low hanging fruit though.